src/doc/using_visit/Quantitative/Query.rst

```diff
@@ -426,7 +426,34 @@ XRay Image
     |                          | the output file is ``outputXXXX.ext``,       |
     |                          | where ``XXXX`` is chosen                     |
     |                          | to be the |br| smallest integer not to       |
-    |                          | overwrite any existing files.                |
+    |                          | overwrite any existing files. As of VisIt_   |
+    |                          | 3.4, |br| it is recommended to use           |
+    |                          | *filename_scheme* in lieu of *family_files*. |
+    +------+-------------------+----------------------------------------------+
+    | *filename_scheme*        | The naming convention for output filenames.  |
+    |                          | This option is available |br| in VisIt_ 3.4, |
+    |                          | and is meant to replace the *family_files*   |
+    |                          | option. If both |br| are provided,           |    
+    |                          | *filename_scheme* will be used.              |
+    +------+-------------------+----------------------------------------------+
+    |      | "none" or 0       | The default. Output filenames will be of the |
+    |      |                   | form ``output.ext``, where |br|              |
+    |      |                   | ``ext`` is the file extension. If the        |
+    |      |                   | filename already exists, VisIt_ will |br|    |
+    |      |                   | overwrite it.                                |
+    +------+-------------------+----------------------------------------------+
+    |      | "family" or 1     | If on, VisIt_ will attempt to family output  |
+    |      |                   | files. Output filenames will be |br| of the  | 
+    |      |                   | form ``output.XXXX.ext``, where ``XXXX`` is  |
```

Comment:
Is the total number of digits to appear fixed or variable? If latter, how is that controlled? If former, what is that number? Examples here show 4 `X` chars but below you show 6. Maybe thats because cycle numbers can tend to go to much larger values than sequence of files?

Reply:
The total number of digits is fixed. Familying, just like before, will always use 4 characters. Cycle info used 6 characters in blueprint filenames so I thought that made sense to replicate here. If someone long ago thought that 6 digits were needed for cycle info, who am I to think differently?

Reply:
This almost makes we wish for a slightly different API for doing this...

* `std::string file_number_scheme = ""` ==> `NONE`
* `std::string file_number_scheme = "%06d$c"` ==> `6 digits w/leading zeros of Cycle`
* `std::string file_number_scheme = "%d$c"` ==> `any number of digits, no leading zeros of Cycle`
* `std::string file_number_scheme = "%03d$i"` ==> `3 digits w/leading zeros of Index`
* `std::string file_number_scheme = "0x%X$i"` ==> `any number of hex digits, no leading zeros of Index`


So, there is just a string value holding a sprintf-style numbering convention with magic `$c` (and friends) for particular datum type to use for the number. This is similar to other parts of VisIt (annotation macros) as well as other parts of Linux (date typesetting for example).

It provides more control over format and there is no enuming involved.

Reactions?

Reply:
You run into danger with indexing. Indexing only occurs when it has to. We can't allow users to turn it off, otherwise they'll overwrite files as the query executes. Indexing used to always be on, so most outputs used to look like `output.00.png`.

Reply:
Doesn't `NONE` mode turn it off? Otherwise, I guess I don't understand.

Reply:
The `filename_scheme` does not affect bin info in filenames. Kind of unintuitive, I know. Bin info in the filenames is caused not by what the user selects, but by a couple factors. See `multipleOutputFiles` on line 123 of the x ray image query file.

Putting bin info in the filename was never something the user could control before. It will be put in if the query thinks it is needed, and skipped if not. The old behavior was to always put it in, although our docs did not reflect that. I thought it would be cleaner if it only happened when it needed to, but if you felt strongly they other way I could change it.

Reply:
Ok, I think I understand. That means that two-digit numbering part of the name is either needed (or its not) and the user should have no control over that. Sure, that makes sense. But, I guess that is a bit orthogonal to the question I was raising about the interface. Instead of `int filenameScheme` with 3 magic numbers (currently)...

why not `std::string filenameScheme` where  `""` means what `NONE` means and `"$Q%F` means to *render* datum `Q` according to (optional) printf-spec `F`. `"$c%06d"` means what `CYCLEFILES` means here and `"$s%04d"` means what `FAMILYFILES` (or maybe `"$i%04d"` where i=*index*) means here but where but where users can then easily control the *number* of digits as well as some of the way it gets formatted into the filename including `"$t%8.4g"` could render a time value instead of cycle or index or (borrowing from [this stuff](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/using_visit/MakingItPretty/Annotations.html?highlight=dbcomment#database-values-in-text-annotations) to some extent.

Reply:
In fact, there is no need for an extra `std::string filenameScheme` data member for this. Just make it handled as part of the `filename` string itself.

I realize what I am proposing here is a departure from what you've already coded. But, what  I propose is consistent with other parts of VisIt in handling things like an index or a cycle number in a string and offers more flexibility to the user.

If you are dead set against it though, let me know and I'll go ahead and approve as is.

Reply:
I am not dead set against it; I think giving users more options for what goes in the filename is a very worthy goal.

However, for this PR, I have endeavored to (mostly) preserve how things have worked in the past, while trying also to standardize filename conventions between the different output types. (For context, prior to these changes, there were many _many_ special cases all over the place for the different output types.) I think that in the interest of keeping changes small and focused, we should save these ideas for future work.

What I will do is take your comments and add them to my ticket #17791 recording ideas for future work on the x ray query. Additionally, if we do opt to depart from the current (very limited) options users have for naming output files, I think that may be worth discussing with others (maybe just a 10-15 minute slice of a special topics meeting). I know that Cyrus has a great understanding of what users need and expect, and Eric originated much of this code and would likely have some good ideas as well. Now that you've had a chance to look at it (and hopefully I've been able to answer some of your questions), I think a good discussion could come of these ideas.

I don't mean to come off as dead set against your suggestions! Sorry if it seems that way :( Your comments have highlighted to me that a lot should be done to bring this code up to date and up to snuff. I think we should continue to think about improvements that can be made here, and as I continue making changes to the query, I should look more to overhaul legacy decisions and logic as opposed to writing code that preserves it.

Reply:
TLDR, please approve as is and I will take note of your ideas for future work on the query. I have every intention of continuing to make it better.